File: homework3/Composition/Composition.kt
package Composition

// WalkMethod 인터페이스 정의
interface WalkMethod {
    fun walk(): Int
}

// WalkToSchool 클래스가 WalkMethod를 구현
class WalkToSchool(private val distance: Int) : WalkMethod {
    override fun walk(): Int {
        println("Walking to school, distance: $distance km")
        return distance
    }
}

// WalkHome 클래스가 WalkMethod를 구현
class WalkHome(private val distance: Int) : WalkMethod {
    override fun walk(): Int {
        println("Walking home, distance: $distance km")
        return distance
    }
}

// Person 클래스
class Person(private val name: String, private val walkMethod: WalkMethod) {
    fun walk() {
        println("$name is walking...")
        walkMethod.walk()
    }
}

// main 함수
fun main() {
    val john = Person("John", WalkToSchool(2)) // WalkToSchool 객체 전달
    john.walk()

    val jane = Person("Jane", WalkHome(3)) // WalkHome 객체 전달
    jane.walk()
}

File: homework3/Composition/Composition.cpp
#include <iostream>
#include <memory>
#include <string>

class WalkMethod {
public:
    virtual ~WalkMethod() = default;
    virtual int walk() const = 0;
};

class WalkToSchool : public WalkMethod {
private:
    int distance;
public:
    WalkToSchool(int distance) : distance(distance) {}
    int walk() const override {
        std::cout << "Walking to school, distance: " << distance << " km" << std::endl;
        return distance;
    }
};

class WalkHome : public WalkMethod {
private:
    int distance;
public:
    WalkHome(int distance) : distance(distance) {}
    int walk() const override {
        std::cout << "Walking home, distance: " << distance << " km" << std::endl;
        return distance;
    }
};

class Person {
private:
    std::string name;
    std::shared_ptr<WalkMethod> walkMethod;
public:
    Person(const std::string& name, std::shared_ptr<WalkMethod> walkMethod)
        : name(name), walkMethod(walkMethod) {}

    void walk() const {
        std::cout << name << " is walking..." << std::endl;
        walkMethod->walk();
    }
};

int main() {
    std::shared_ptr<WalkMethod> walkToSchool = std::make_shared<WalkToSchool>(2);
    Person john("John", walkToSchool);
    john.walk();

    std::shared_ptr<WalkMethod> walkHome = std::make_shared<WalkHome>(3);
    Person jane("Jane", walkHome);
    jane.walk();

    return 0;
}